neurobooth_os/iout/stim_param_reader.py

```diff
@@ -0,0 +1,166 @@
+from os import environ, path
+
+from pydantic import BaseModel, ConfigDict, NonNegativeFloat, NonNegativeInt, Field
+from typing import Optional, List, Callable
+import os
+import yaml
+
+"""
+Loads yaml files containing task/stimulus/instruction parameters and validates them.
+
+To allow variety in the representation of the task parameters, the yaml reading process proceeds in 
+two phases. The first phase reads enough to reach the top-level element arg_parser, e.g.:
+arg_parser: iout.stim_param_reader.py::EyeTrackerStimArgs()
+
+The arg_parser points to a class that inherits from StimulusArgs (which itself inherits from 
+the pydantic BaseModel class). This class handles the parsing of the yaml file in its entirety.
+
+Parsers for all the standard stimulus yaml files are found in this module.   
+"""
+
+
+class InstructionArgs(BaseModel):
+    """
+        Arguments controlling psychopy instructions
+    """
+    instruction_text: Optional[str]
+    instruction_filetype: Optional[str]
+    instruction_file: Optional[str]
+
+
+class StimulusArgs(BaseModel):
+    """
+    Stimulus arguments common to all Psychopy tasks
+    """
+    stimulus_id: str = Field(min_length=1, max_length=255)
+    stimulus_description: str = Field(min_length=1, max_length=255)
+    prompt: Optional[bool] = True
+    arg_parser: str = Field(min_length=1, max_length=255)
+    num_iterations: NonNegativeInt
+    duration: Optional[NonNegativeFloat] = None
+    stimulus_file_type: str = Field(min_length=1, max_length=255)
+    stimulus_file: str = Field(min_length=1, max_length=255)
+
+    task_repeatable_by_subject: Optional[bool] = True
+
+    model_config = ConfigDict(extra='forbid', frozen=True)
+
+
+class TaskArgs(BaseModel):
+    """
+    Task function objects and associated arguments
+    """
+    task_id: str = Field(min_length=1, max_length=255)
+    task_constructor_callable: Callable  # callable of constructor for a Task
+    stim_args: StimulusArgs
+    instr_args: Optional[InstructionArgs] = None
+    task_instance: Optional[object] = None  # created by client code from above callable
+
+    class Config:
+        arbitrary_types_allowed = True
+
+
+class EyeTrackerStimArgs(StimulusArgs):
```

Comment:
The location is specified in the yaml file so anyone adding a new task can put them wherever they like.  For the existing tasks I think it's easier to have them in one place because the inheritance relationships make it easier to read them that way (and it loosely mirrors the Task hierarchy organization).